src/hotspot/share/cds/heapShared.cpp

```diff
@@ -836,6 +971,51 @@ void HeapShared::resolve_classes_for_subgraphs(JavaThread* current, ArchivableSt
   }
 }
 
+// The main purpose of this call is to initialize any classes that are reachable
+// from the archived Java mirrors that belong to the <class_loader>.
+//
+// For example, if this enum class is initialized at AOT cache assembly time:
+//
+//    enum Fruit {
+//       APPLE, ORANGE, BANANA;
+//       static final Set<Fruit> HAVE_SEEDS = new HashSet<>(Arrays.asList(APPLE, ORANGE));
+//   }
+//
+// the pre-inited mirror of Fruit references HashSet, which should be initialized
+// before any Java code can access the Fruit class.
+//
+// The set of classes that are required to be initialized for the archived
+// java mirrors are recorded in _runtime_default_subgraph_info (which probably
+// needs a better name).
+void HeapShared::init_classes_reachable_from_archived_mirrors(Handle class_loader, TRAPS) {
+  if (!ArchiveHeapLoader::is_in_use()) {
+    return;
+  }
+
+  assert( _runtime_default_subgraph_info != nullptr, "must be");
+  Array<Klass*>* klasses = _runtime_default_subgraph_info->subgraph_object_klasses();
```

Comment:
Couple of questions here:
1. Does `_runtime_default_subgraph_info` only hold archived mirrors? 

2. If we are init-ing classes required for archived mirrors here, why do we need special aot-initialization for `PrimitiveClassDescImpl`, `ReferenceClassDescImpl` and `ConstantDescs` in AOTClassInitializer::can_archive_initialized_mirror?

Reply:
> Does _runtime_default_subgraph_info only hold information about archived mirrors?

The default graph records classes of objects created in this step:

```
void HeapShared::copy_objects() {
  assert(HeapShared::can_write(), "must be");

  copy_interned_strings(); // << here
  copy_special_objects();  // << and here
```

So apart from mirrors, it also records the class of String and 6 exception instances archived by `HeapShared::archive_exception_instance`


> If we are init-ing classes required for archived mirrors here, why do we need special aot-initialization for PrimitiveClassDescImpl, ReferenceClassDescImpl and ConstantDescs in AOTClassInitializer::can_archive_initialized_mirror?

I've added the following comments into `AOTClassInitializer::can_archive_initialized_mirror()`

```
  } else if (ik->is_initialized() &&
             (ik->name()->equals("jdk/internal/constant/PrimitiveClassDescImpl") ||
              ik->name()->equals("jdk/internal/constant/ReferenceClassDescImpl") ||
              ik->name()->equals("java/lang/constant/ConstantDescs"))) {
    // The above 3 classes are special cases needed to support the aot-caching of
    // java.lang.invoke.MethodType instances:
    // - MethodType points to sun.invoke.util.Wrapper enums
    // - The Wrapper enums point to static final fields in the above 3 classes.
    //   E.g., ConstantDescs.CD_Boolean.
    // - If we re-run the <clinit> of these 3 classes again during the production
    //   run, ConstantDescs.CD_Boolean will get a new value that has a different
    //   object identity than the value referenced the the Wrapper enums.
    // - However, Wrapper requires object identity (it allows the use of == to
    //   test the equality of ClassDesc, etc).
    // Therefore, we must preserve the static fields of these 3 classes from
    // the assembly phase.
    return true;
```

> The comment seems to indicate this function should initialize the classes that are referenced by archived mirrors of aot-initialized classes. But there is no check for aot-initialized class in the body.

I amended the comments above `HeapShared::init_classes_reachable_from_archived_mirrors()` to say:

```
//    enum Fruit {
//       APPLE, ORANGE, BANANA;
//       static final Set<Fruit> HAVE_SEEDS = new HashSet<>(Arrays.asList(APPLE, ORANGE));
//   }
//
// the pre-inited mirror of Fruit references HashSet, which should be initialized
// before any Java code can access the Fruit class. Note that HashSet itself doesn't
// necessary need to be an aot-initialized class.
```

That's why there's no check for `k` to be aot-initialized.

This is necessary because the `<clinit>` of `k` could have environment dependencies so it cannot be aot-initialized.

Reply:
I was actually referring to the missing aot-initialized check for the Fruit class.
As it stands, this method initializes the classes required by the archive mirrors as the _runtime_default_subgraph_info has all the archived mirrors. But not all classes that have archived mirror are aot-initialized. And from the Fruit class example in the comment it seems this method should only be initializing the classes that are required by archived mirrors of _aot-initialized classes_:

```
// For example, if this enum class is initialized at AOT cache assembly time:
//
//    enum Fruit {
//       APPLE, ORANGE, BANANA;
//       static final Set<Fruit> HAVE_SEEDS = new HashSet<>(Arrays.asList(APPLE, ORANGE));
//   }
//
// the pre-inited mirror of Fruit references HashSet, which should be initialized
// before any Java code can access the Fruit class.
```

So based on the comment there should be a way to identify the subgraph_object_klasses of only the aot-initialized classes and initialize only those classes.
Am I reading this wrong?

Reply:
> I was actually referring to the missing aot-initialized check for the Fruit class. As it stands, this method initializes the classes required by the archive mirrors as the _runtime_default_subgraph_info has all the archived mirrors.

`_runtime_default_subgraph_info` is not recording the mirrors. It records all the classes of all the objects that can are reachable from the archived mirrors.

For example, if the following three classes are aot-initialized:

```
class A { static Object foo = new X(); }
class B { static Object foo = new Y(); }
class C { static Object foo = new Y(); }
```

`_runtime_default_subgraph_info`  records `X` and `Y`. It doesn't record `A`, `B`, or `C`.

> But not all classes that have archived mirror are aot-initialized. 

If a class is not AOT initialized, its mirror is filled with zeros (plus a few native pointers) so the mirror doesn't point to any object.

Reply:
Expanding on the above example, lets say A is aot- initialized, but B and C are not.
So this function should initialize only X not Y, is that correct? If so, then how does it prevent initialization of Y? It iterates through all the subgraph_object_klasses which includes both X and Y.

Reply:
The `subgraph_object_klasses` are built during assembly phase when each mirror is added to the cache. Note that we don't add the "real" mirror of the classes, but we add the scratch mirror:

```
void HeapShared::archive_java_mirrors() {
    ...
    oop m = scratch_java_mirror(orig_k);
    if (m != nullptr) {
      Klass* buffered_k = ArchiveBuilder::get_buffered_klass(orig_k);
      bool success = archive_reachable_objects_from(1, _default_subgraph_info, m);
```

So the scratch mirrors of `B` and `C` will be empty when they are being archived.

Because `A` is marked as aot-initialized, we copy the fields of the "real" mirror of `A` into its scratch mirror. See `HeapShared::copy_aot_initialized_mirror()`. That's why we are able to add `X` into the subgraph_object_klasses (when `A`'s scratch mirror is scanned inside `archive_reachable_objects_from()`)

Reply:
Oh I see! I was missing the point that scratch java mirrors are empty. So their clinit dependencies don't get recorded in subgraph_object_klasses. I should have seen that before. Sorry for dragging it.

The point that is bothering me is this code relies on the assumption that the default subgraph mainly consists of mirror objects, some of which may be aot-initialized, while others are emply. If in future we added other objects (not the mirror objects) to the graph, then this assumption would fail. But we don't have any check to protect against this.
Can we create a separate subgraph for aot-initialized mirror objects to record their clinit dependencies and process them here? That would make the code explicit in what it intends to do.

Reply:
@iklam are you planning to address this? Otherwise the patch looks good to me.

Reply:
Yes, I plan to to address this.

I think the name `init_classes_reachable_from_archived_mirrors` is confusing, as it also initializes classes that may not be reachable from archived mirrors. I will try to rename this function and related data structures to make the meaning more obvious.

I'd like to keep the logic the same, as I still need to do the initialization in 4 steps (boo1/boot2/platform/app) for classes listed under `_runtime_default_subgraph_info`. If I split up `_runtime_default_subgraph_info` into two separate parts the code will just become more verbose.

Reply:
I renamed the function to `HeapShared::init_classes_for_special_subgraph()`. I also changed other `_default_subgraph` names to `special_subgraph`.